url43a0865bb0_kevinjdolan_darwin_tgz-pJ8-tui_DriverJ8/src/darwInvest/data/Ticker.java

```diff
@@ -0,0 +1,373 @@
+	}
+	
+	/**
+	 * Normalize the data to be between the min and max
+	 * for the particular data field
+	 * @param which the index of the data field
+	 * @param data  the value of the data
+	 * @return		a number (preferably) [0,1]
+	 */
+	public double normalizeData(int which, double data) {
+		double diff = data - minData[which];
+		double range = maxData[which] - minData[which];
+		return diff / range;
+	}
+	
+	/**
+	 * Get a list of all news stories in a given time range
+	 * @param firstTime the beginning of the time range
+	 * @param lastTime	the end of the time range
+	 * @return			the list of news events
+	 */
+	public List<NewsEvent> getNews(long firstTime, long lastTime) {
+		Collection<List<NewsEvent>> collection = 
+			(news.subMap(firstTime, true, lastTime,true)).values();
+		List<NewsEvent> list = new ArrayList<NewsEvent>();
+		for(List<NewsEvent> entry : collection)
+			for(NewsEvent event : entry)
+				list.add(event);
+		return list;
+	}
+	
+	/**
+	 * @return the first time of this data set (considering only price information)
+	 */
+	public long getFirstTime() {
+		return data.firstKey();
+	}
+	
+	/**
+	 * @return the last time of this data set (considering only price information)
+	 */
+	public long getLastTime() {
+		return data.lastKey();
+	}
+	
+	/**
+	 * @return the first time of the news data set
+	 */
+	public long getFirstNews() {
+		return news.firstKey();
+	}
+	
+	/**
+	 * @return the last time of the news data set
+	 */
+	public long getLastNews() {
+		return news.lastKey();
+	}
+	
+	/**
+	 * Return all times between the given range
+	 * @param begin the beginning of the range
+	 * @param end   the end of the range
+	 * @return      an ordered set of time values
+	 */
+	public Set<Long> getTimesBetween(long begin, long end) {
+		return data.subMap(begin, end).keySet();
+	}
+	
+	/**
+	 * Get the lowest time greater than the input time
+	 * @param time the time to look for
+	 * @return     the ceiling time
+	 */
+	public long ceilingTime(long time) {
+		return data.ceilingKey(time);
+	}
+	
+	/**
+	 * Get the greatest time less than the input time
+	 * @param time the time to look for
+	 * @return     the floor time
+	 */
+	public long floorTime(long time) {
+		return data.floorKey(time);
+	}
+
+	public String lookupGooglecompanyid() throws IOException {
+		try {
+			URL google = new URL("http://www.google.com/finance?q=NASDAQ:" + symbol);
+			BufferedReader in = new BufferedReader(new InputStreamReader(google.openStream()));
+			Scanner s = new Scanner(in);
+			s.useDelimiter("[\n\r]+");
+			String companyid = null;
+			while(s.hasNext()) {
+				String next = s.next();
+				if(next.contains("var _companyId")) {
+					String extractor = next.substring(17);
+					companyid = extractor.substring(0,(extractor.indexOf(';')));
+					break;
+				}
+			}
+			if (s != null) {
+				s.close();
+			}
+			if (in != null) {
+				in.close();
+			}
+			this.googlecompanyid = companyid;
+			return companyid;
+		}
+		catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)